main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Differ.scala

```diff
@@ -26,15 +26,66 @@ object Differ {
     case None =>
       // Case 1: No old AST. No difference.
       DeltaContext(Nil)
-    case Some(oldAst) => {
+    case Some(oldAst) =>
       // Case 2: We have an oldAst and a newAst. Compute their difference.
 
       // TODO: Add some comments and introduce helper functions.
-      val newDefs = (newAst.defs.keySet -- oldAst.defs.keySet).toList.map(sym => Delta.AddDef(sym, getCurrentTimestamp))
+      val newDefs = calcDefDiff(oldAst, newAst)
       val newEnums = (newAst.enums.keySet -- oldAst.enums.keySet).toList.map(sym => Delta.AddEnum(sym, getCurrentTimestamp))
 
       DeltaContext(newDefs ++ newEnums)
-    }
+  }
+
+  /**
+    * Computes the difference between oldAst and newAst for defs
+    *
+    * @param oldAst the old AST.
+    * @param newAst the new AST.
+    * @return       a list of changes as Delta's.
+    */
+  private def calcDefDiff(oldAst: TypedAst.Root, newAst: TypedAst.Root): List[Delta] = {
+    // Find all newly added defs
+    val newlyAddedDefs = (newAst.defs.keySet -- oldAst.defs.keySet).toList.map(sym => Delta.AddDef(sym, getCurrentTimestamp))
+    // Check if any of the oldDefs has changed implementation
+    val changedDefs =
+      newAst.defs.keySet.toList.flatMap(sym =>
+        oldAst.defs.get(sym) match {
+          case Some(oldDef) =>
+            newAst.defs.get(sym) match {
+              case Some(newDef) =>
+                // Case 1: We have the same sym in both new and old AST. Check if the implementation has changed
+                if (compareDefs(oldDef, newDef)) {
+                  // Decl has changed
+                  Some(Delta.AddDef(sym, getCurrentTimestamp))
+                } else {
+                  // Decl hasn't changed
+                  None
+                }
+              case None =>
+                // The def has been deleted
+                None
+            }
+          case None =>
+            // It's a newly added def
+            None
+        })
+    newlyAddedDefs ++ changedDefs
+  }
+
+  /**
+    * Compares two defs in terms of change
+    *
+    * @param oldDef the old def.
+    * @param newDef the new def.
+    * @return       true if the def has changed, false otherwise.
+    */
+  private def compareDefs(oldDef: TypedAst.Def, newDef: TypedAst.Def): Boolean = {
```

Comment:
I'm only able to capture changes in terms of location in the file. However, this doesn't necessarily catch all changes.